File: sdk/typescript/bedrock-cost-wrapper.ts
/**
 * Bedrock Cost Explorer — TypeScript instrumentation wrapper.
 *
 * Wraps @aws-sdk/client-bedrock-runtime calls (InvokeModel, Converse)
 * to capture token usage and attribution metadata, then ships the event
 * to the Cost Explorer ingest API.
 *
 * Usage:
 *
 *   import { BedrockCostWrapper } from "./bedrock-cost-wrapper";
 *
 *   const wrapper = new BedrockCostWrapper({
 *     ingestApiUrl: "https://xxx.execute-api.us-east-1.amazonaws.com/v1/events",
 *     apiKey: process.env.BEDROCK_COST_API_KEY!,
 *     applicationId: "my-search-service",
 *   });
 *
 *   const bedrock = wrapper.client({ region: "us-east-1" });
 *
 *   const response = await bedrock.converse({
 *     modelId: "anthropic.claude-sonnet-4-5",
 *     messages: [{ role: "user", content: [{ text: "Hello" }] }],
 *     // Attribution — stripped before the SDK call
 *     _userId: "user_abc123",
 *     _agentId: undefined,
 *     _sessionId: "sess_xyz",
 *   });
 */

import {
  BedrockRuntimeClient,
  BedrockRuntimeClientConfig,
  ConverseCommand,
  ConverseCommandInput,
  ConverseCommandOutput,
  InvokeModelCommand,
  InvokeModelCommandInput,
  InvokeModelCommandOutput,
} from "@aws-sdk/client-bedrock-runtime";
import { STSClient, GetCallerIdentityCommand } from "@aws-sdk/client-sts";
import { randomUUID } from "crypto";
import { fetch } from "undici"; // Node 18+ built-in fetch fallback

// ─────────────────────────────────────────────
// Types
// ─────────────────────────────────────────────

export interface WrapperConfig {
  ingestApiUrl: string;
  apiKey: string;
  applicationId: string;
  accountId?: string;
  /** Emit events asynchronously (default: true). Set false in tests. */
  emitAsync?: boolean;
  maxRetries?: number;
  timeoutMs?: number;
}

export interface AttributionOptions {
  _userId?: string | null;
  _agentId?: string | null;
  _sessionId?: string | null;
}

export interface InvocationEvent {
  event_id: string;
  schema_version: number;
  timestamp: string;
  region: string;
  account_id: string;
  model_id: string;
  invocation_type: "invoke_model" | "converse" | "agent_step";
  user_id: string | null;
  agent_id: string | null;
  session_id: string | null;
  application_id: string;
  request_id: string | null;
  input_tokens: number;
  output_tokens: number;
  cache_read_tokens: number;
  cache_write_tokens: number;
  latency_ms: number;
  status: "success" | "error" | "throttled";
  error_code: string | null;
  source: "wrapper" | "cloudwatch_backfill";
}

// ─────────────────────────────────────────────
// BedrockCostWrapper
// ─────────────────────────────────────────────

export class BedrockCostWrapper {
  private readonly config: Required<WrapperConfig>;
  private _accountId: string | null = null;
  private _accountIdPromise: Promise<string> | null = null;

  constructor(config: WrapperConfig) {
    this.config = {
      accountId: undefined as unknown as string,
      emitAsync: true,
      maxRetries: 2,
      timeoutMs: 5000,
      ...config,
    };
  }

  private async getAccountId(): Promise<string> {
    if (this.config.accountId) return this.config.accountId;
    if (this._accountId) return this._accountId;
    if (!this._accountIdPromise) {
      this._accountIdPromise = (async () => {
        try {
          const sts = new STSClient({});
          const resp = await sts.send(new GetCallerIdentityCommand({}));
          this._accountId = resp.Account ?? "unknown";
          return this._accountId;
        } catch {
          this._accountId = "unknown";
          return this._accountId;
        }
      })();
    }
    return this._accountIdPromise;
  }

  /** Create an instrumented BedrockRuntime client. */
  client(clientConfig: BedrockRuntimeClientConfig = {}): WrappedBedrockClient {
    const raw = new BedrockRuntimeClient(clientConfig);
    const region =
      (typeof clientConfig.region === "string" ? clientConfig.region : null) ??
      process.env.AWS_DEFAULT_REGION ??
      "us-east-1";
    return new WrappedBedrockClient(raw, this, region);
  }

  /** Wrap an existing BedrockRuntimeClient. */
  wrap(client: BedrockRuntimeClient, region?: string): WrappedBedrockClient {
    const r = region ?? process.env.AWS_DEFAULT_REGION ?? "us-east-1";
    return new WrappedBedrockClient(client, this, r);
  }

  async emitEvent(event: InvocationEvent): Promise<void> {
    // Fill in account_id async (cached after first call)
    event.account_id = await this.getAccountId();

    if (this.config.emitAsync) {
      // Fire-and-forget; errors logged, never thrown
      this.sendEvent(event).catch((err) => {
        console.error(
          `[BedrockCostWrapper] Failed to emit event ${event.event_id}:`,
          err
        );
      });
    } else {
      await this.sendEvent(event);
    }
  }

  private async sendEvent(
    event: InvocationEvent,
    attempt = 0
  ): Promise<void> {
    const controller = new AbortController();
    const timer = setTimeout(
      () => controller.abort(),
      this.config.timeoutMs
    );
    try {
      const resp = await fetch(this.config.ingestApiUrl, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "X-Api-Key": this.config.apiKey,
        },
        body: JSON.stringify(event),
        signal: controller.signal,
      });
      if (!resp.ok) {
        const body = await resp.text();
        console.warn(
          `[BedrockCostWrapper] Ingest API ${resp.status}: ${body}`
        );
      }
    } catch (err) {
      if (attempt < this.config.maxRetries) {
        const backoff = 200 * Math.pow(2, attempt);
        await new Promise((r) => setTimeout(r, backoff));
        return this.sendEvent(event, attempt + 1);
      }
      throw err;
    } finally {
      clearTimeout(timer);
    }
  }

  get applicationId(): string {
    return this.config.applicationId;
  }
}

// ─────────────────────────────────────────────
// WrappedBedrockClient
// ─────────────────────────────────────────────

export class WrappedBedrockClient {
  constructor(
    private readonly _client: BedrockRuntimeClient,
    private readonly _wrapper: BedrockCostWrapper,
    private readonly _region: string
  ) {}

  private extractAttribution(input: Record<string, unknown>): AttributionOptions {
    const attr: AttributionOptions = {
      _userId: (input["_userId"] as string) ?? null,
      _agentId: (input["_agentId"] as string) ?? null,
      _sessionId: (input["_sessionId"] as string) ?? null,
    };
    delete input["_userId"];
    delete input["_agentId"];
    delete input["_sessionId"];
    return attr;
  }

  async converse(
    input: ConverseCommandInput & AttributionOptions
  ): Promise<ConverseCommandOutput> {
    const mutableInput = { ...input } as Record<string, unknown>;
    const attribution = this.extractAttribution(mutableInput);
    const modelId = (mutableInput["modelId"] as string) ?? "unknown";

    const start = performance.now();
    let status: "success" | "error" | "throttled" = "success";
    let errorCode: string | null = null;
    let response: ConverseCommandOutput | null = null;

    try {
      response = await this._client.send(
        new ConverseCommand(mutableInput as ConverseCommandInput)
      );
      return response;
    } catch (err: unknown) {
      status = "error";
      const awsErr = err as { name?: string };
      errorCode = awsErr?.name ?? "UnknownError";
      if (errorCode === "ThrottlingException") status = "throttled";
      throw err;
    } finally {
      const latencyMs = performance.now() - start;
      const usage = response?.usage ?? {};

      const event: InvocationEvent = {
        event_id: randomUUID(),
        schema_version: 1,
        timestamp: new Date().toISOString(),
        region: this._region,
        account_id: "pending", // filled by emitEvent
        model_id: modelId,
        invocation_type: "converse",
        user_id: attribution._userId ?? null,
        agent_id: attribution._agentId ?? null,
        session_id: attribution._sessionId ?? null,
        application_id: this._wrapper.applicationId,
        request_id:
          (response?.$metadata?.requestId as string | undefined) ?? null,
        input_tokens: usage.inputTokens ?? 0,
        output_tokens: usage.outputTokens ?? 0,
        cache_read_tokens: (usage as Record<string, number>).cacheReadInputTokens ?? 0,
        cache_write_tokens: (usage as Record<string, number>).cacheWriteInputTokens ?? 0,
        latency_ms: Math.round(latencyMs),
        status,
        error_code: errorCode,
        source: "wrapper",
      };

      // Don't await — emit fires in background
      this._wrapper.emitEvent(event);
    }
  }

  async invokeModel(
    input: InvokeModelCommandInput & AttributionOptions
  ): Promise<InvokeModelCommandOutput> {
    const mutableInput = { ...input } as Record<string, unknown>;
    const attribution = this.extractAttribution(mutableInput);
    const modelId = (mutableInput["modelId"] as string) ?? "unknown";

    const start = performance.now();
    let status: "success" | "error" | "throttled" = "success";
    let errorCode: string | null = null;
    let response: InvokeModelCommandOutput | null = null;

    try {
      response = await this._client.send(
        new InvokeModelCommand(mutableInput as InvokeModelCommandInput)
      );
      return response;
    } catch (err: unknown) {
      status = "error";
      const awsErr = err as { name?: string };
      errorCode = awsErr?.name ?? "UnknownError";
      if (errorCode === "ThrottlingException") status = "throttled";
      throw err;
    } finally {
      const latencyMs = performance.now() - start;

      // Parse token counts from response body (Claude models)
      let inputTokens = 0;
      let outputTokens = 0;
      let cacheRead = 0;
      let cacheWrite = 0;
      if (response?.body) {
        try {
          const body = JSON.parse(
            new TextDecoder().decode(response.body as Uint8Array)
          );
          const usage = body?.usage ?? {};
          inputTokens = usage.input_tokens ?? 0;
          outputTokens = usage.output_tokens ?? 0;
          cacheRead = usage.cache_read_input_tokens ?? 0;
          cacheWrite = usage.cache_creation_input_tokens ?? 0;
        } catch {
          // Non-JSON or non-Claude model response — tokens remain 0
        }
      }

      const event: InvocationEvent = {
        event_id: randomUUID(),
        schema_version: 1,
        timestamp: new Date().toISOString(),
        region: this._region,
        account_id: "pending",
        model_id: modelId,
        invocation_type: "invoke_model",
        user_id: attribution._userId ?? null,
        agent_id: attribution._agentId ?? null,
        session_id: attribution._sessionId ?? null,
        application_id: this._wrapper.applicationId,
        request_id:
          (response?.$metadata?.requestId as string | undefined) ?? null,
        input_tokens: inputTokens,
        output_tokens: outputTokens,
        cache_read_tokens: cacheRead,
        cache_write_tokens: cacheWrite,
        latency_ms: Math.round(latencyMs),
        status,
        error_code: errorCode,
        source: "wrapper",
      };

      this._wrapper.emitEvent(event);
    }
  }

  /** Expose the underlying client for operations we don't instrument. */
  get rawClient(): BedrockRuntimeClient {
    return this._client;
  }
}

File: sdk/python/bedrock_cost_wrapper.py
"""
Bedrock Cost Explorer — Python instrumentation wrapper.

Wraps boto3 Bedrock runtime calls (invoke_model, converse) to capture
token usage, cost attribution metadata, and invocation context, then
ships the event to the Cost Explorer ingest API.

Usage:

    from bedrock_cost_wrapper import BedrockCostWrapper

    wrapper = BedrockCostWrapper(
        ingest_api_url="https://xxx.execute-api.us-east-1.amazonaws.com/v1/events",
        api_key="your-api-key",
        application_id="my-search-service",
    )

    # Option A: create a pre-configured boto3 client
    bedrock = wrapper.client(region_name="us-east-1")
    response = bedrock.converse(
        modelId="anthropic.claude-sonnet-4-5",
        messages=[{"role": "user", "content": [{"text": "Hello"}]}],
        # Pass attribution via additional keyword arguments — stripped before the call
        _user_id="user_abc123",
        _agent_id=None,
        _session_id="sess_xyz",
    )

    # Option B: wrap an existing client
    import boto3
    existing_client = boto3.client("bedrock-runtime", region_name="us-east-1")
    bedrock = wrapper.wrap(existing_client)

Design intent:
- The wrapper intercepts the RESPONSE to read token counts.
  It does NOT proxy or inspect the request payload (privacy-first).
- Attribution fields are passed via underscore-prefixed kwargs (_user_id etc.)
  so they don't pollute the boto3 call signature.
- Events are emitted asynchronously (background thread) to add zero latency
  to the Bedrock call itself.
- Failures in event emission are logged but never raise — the wrapper must
  never break the application.
"""

import json
import logging
import os
import threading
import time
import uuid
from datetime import datetime, timezone
from functools import wraps
from typing import Any, Callable, Dict, Optional
from urllib import request, error as urllib_error

import boto3
import botocore

logger = logging.getLogger(__name__)


class BedrockCostWrapper:
    """
    Instrumentation wrapper for boto3 Bedrock Runtime clients.

    Parameters
    ----------
    ingest_api_url : str
        URL of the Cost Explorer ingest endpoint (POST /events).
    api_key : str
        API key for the ingest endpoint.
    application_id : str
        Logical name of the application using Bedrock.
    account_id : str, optional
        AWS account ID. Auto-detected from STS if not provided.
    emit_async : bool
        If True (default), emit events in a background thread.
        Set to False in tests or Lambda environments where the process
        may exit before the background thread completes.
    max_retries : int
        Number of times to retry failed event emission. Default: 2.
    timeout_seconds : float
        HTTP timeout for event emission. Default: 5.0.
    """

    ATTRIBUTION_KWARGS = {"_user_id", "_agent_id", "_session_id"}

    def __init__(
        self,
        ingest_api_url: str,
        api_key: str,
        application_id: str,
        account_id: Optional[str] = None,
        emit_async: bool = True,
        max_retries: int = 2,
        timeout_seconds: float = 5.0,
    ):
        self.ingest_api_url = ingest_api_url.rstrip("/")
        self.api_key = api_key
        self.application_id = application_id
        self._account_id = account_id
        self.emit_async = emit_async
        self.max_retries = max_retries
        self.timeout_seconds = timeout_seconds
        self._account_id_lock = threading.Lock()

    def _get_account_id(self) -> str:
        """Lazy-fetch account ID from STS (cached after first call)."""
        if self._account_id:
            return self._account_id
        with self._account_id_lock:
            if not self._account_id:
                try:
                    sts = boto3.client("sts")
                    self._account_id = sts.get_caller_identity()["Account"]
                except Exception as exc:
                    logger.warning("Could not fetch account_id from STS: %s", exc)
                    self._account_id = "unknown"
        return self._account_id

    def client(self, **kwargs) -> "WrappedBedrockClient":
        """
        Create a new boto3 bedrock-runtime client, wrapped for instrumentation.

        All kwargs are passed through to boto3.client().
        """
        region = kwargs.get("region_name", os.environ.get("AWS_DEFAULT_REGION", "us-east-1"))
        raw_client = boto3.client("bedrock-runtime", **kwargs)
        return WrappedBedrockClient(raw_client, self, region)

    def wrap(self, bedrock_client: Any) -> "WrappedBedrockClient":
        """Wrap an existing boto3 bedrock-runtime client."""
        region = bedrock_client.meta.region_name
        return WrappedBedrockClient(bedrock_client, self, region)

    def emit_event(self, event: dict):
        """
        Emit an invocation event to the ingest API.
        Called from WrappedBedrockClient after each invocation.

        NEVER raises — emission failures are always logged and swallowed
        so they cannot break the caller's Bedrock call.
        """
        if self.emit_async:
            t = threading.Thread(target=self._send_event, args=(event,), daemon=True)
            t.start()
        else:
            try:
                self._send_event(event)
            except Exception as exc:
                logger.error(
                    "Failed to emit event %s (sync): %s",
                    event.get("event_id"),
                    exc,
                )

    def _send_event(self, event: dict):
        """HTTP POST with retry. Failures are logged, never raised."""
        payload = json.dumps(event, default=str).encode("utf-8")
        headers = {
            "Content-Type": "application/json",
            "X-Api-Key": self.api_key,
        }

        for attempt in range(self.max_retries + 1):
            try:
                req = request.Request(
                    self.ingest_api_url,
                    data=payload,
                    headers=headers,
                    method="POST",
                )
                with request.urlopen(req, timeout=self.timeout_seconds) as resp:
                    if resp.status not in (200, 201, 202):
                        body = resp.read().decode("utf-8", errors="replace")
                        logger.warning(
                            "Ingest API returned %d: %s (event_id=%s)",
                            resp.status, body, event.get("event_id"),
                        )
                    else:
                        logger.debug("Event %s emitted successfully", event.get("event_id"))
                    return
            except urllib_error.URLError as exc:
                if attempt < self.max_retries:
                    backoff = 0.2 * (2 ** attempt)
                    logger.debug(
                        "Emit attempt %d failed (%s) — retrying in %.1fs",
                        attempt + 1, exc, backoff,
                    )
                    time.sleep(backoff)
                else:
                    logger.error(
                        "Failed to emit event %s after %d attempts: %s",
                        event.get("event_id"), self.max_retries + 1, exc,
                    )
            except Exception as exc:
                logger.error("Unexpected error emitting event %s: %s", event.get("event_id"), exc)
                return


class WrappedBedrockClient:
    """
    Thin wrapper around a boto3 bedrock-runtime client.

    Intercepts invoke_model and converse calls, extracts token usage
    from the response, and emits instrumentation events.

    Attribution is passed as underscore-prefixed kwargs:
        _user_id, _agent_id, _session_id

    These are stripped before the call reaches boto3.
    """

    def __init__(self, client: Any, wrapper: BedrockCostWrapper, region: str):
        self._client = client
        self._wrapper = wrapper
        self._region = region

    def _extract_attribution(self, kwargs: dict) -> dict:
        """Pop attribution kwargs and return them."""
        return {
            "user_id": kwargs.pop("_user_id", None),
            "agent_id": kwargs.pop("_agent_id", None),
            "session_id": kwargs.pop("_session_id", None),
        }

    def _build_event(
        self,
        invocation_type: str,
        model_id: str,
        attribution: dict,
        response: dict,
        latency_ms: float,
        status: str,
        error_code: Optional[str] = None,
        request_id: Optional[str] = None,
    ) -> dict:
        """Build the canonical invocation event from a Bedrock response."""

        # Extract token counts — location differs by API
        usage = {}
        if invocation_type == "converse":
            usage = response.get("usage", {})
            input_tokens = usage.get("inputTokens", 0)
            output_tokens = usage.get("outputTokens", 0)
            cache_read = usage.get("cacheReadInputTokens", 0)
            cache_write = usage.get("cacheWriteInputTokens", 0)
        elif invocation_type == "invoke_model":
            # Claude models return usage in the response body JSON
            body = response.get("body", b"")
            if hasattr(body, "read"):
                body = body.read()
            try:
                parsed = json.loads(body)
                usage = parsed.get("usage", {})
                input_tokens = usage.get("input_tokens", 0)
                output_tokens = usage.get("output_tokens", 0)
                cache_read = usage.get("cache_read_input_tokens", 0)
                cache_write = usage.get("cache_creation_input_tokens", 0)
            except (json.JSONDecodeError, Exception):
                input_tokens = output_tokens = cache_read = cache_write = 0
        else:
            input_tokens = output_tokens = cache_read = cache_write = 0

        return {
            "event_id": str(uuid.uuid4()),
            "schema_version": 1,
            "timestamp": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
            "region": self._region,
            "account_id": self._wrapper._get_account_id(),
            "model_id": model_id,
            "invocation_type": invocation_type,
            "user_id": attribution.get("user_id"),
            "agent_id": attribution.get("agent_id"),
            "session_id": attribution.get("session_id"),
            "application_id": self._wrapper.application_id,
            "request_id": request_id,
            "input_tokens": input_tokens,
            "output_tokens": output_tokens,
            "cache_read_tokens": cache_read,
            "cache_write_tokens": cache_write,
            "latency_ms": int(latency_ms),
            "status": status,
            "error_code": error_code,
            "source": "wrapper",
        }

    def converse(self, **kwargs) -> dict:
        """
        Instrumented wrapper for bedrock-runtime.converse().

        Attribution kwargs: _user_id, _agent_id, _session_id
        All other kwargs are passed through to the boto3 client.
        """
        attribution = self._extract_attribution(kwargs)
        model_id = kwargs.get("modelId", "unknown")

        start = time.monotonic()
        status = "success"
        error_code = None
        response = {}

        try:
            response = self._client.converse(**kwargs)
            return response
        except botocore.exceptions.ClientError as exc:
            status = "error"
            error_code = exc.response.get("Error", {}).get("Code", "UnknownError")
            if error_code == "ThrottlingException":
                status = "throttled"
            raise
        finally:
            latency_ms = (time.monotonic() - start) * 1000
            request_id = (
                response.get("ResponseMetadata", {}).get("RequestId")
                if response
                else None
            )
            event = self._build_event(
                invocation_type="converse",
                model_id=model_id,
                attribution=attribution,
                response=response,
                latency_ms=latency_ms,
                status=status,
                error_code=error_code,
                request_id=request_id,
            )
            self._wrapper.emit_event(event)

    def invoke_model(self, **kwargs) -> dict:
        """
        Instrumented wrapper for bedrock-runtime.invoke_model().

        Attribution kwargs: _user_id, _agent_id, _session_id
        All other kwargs are passed through to the boto3 client.
        """
        attribution = self._extract_attribution(kwargs)
        model_id = kwargs.get("modelId", "unknown")

        start = time.monotonic()
        status = "success"
        error_code = None
        response = {}

        try:
            response = self._client.invoke_model(**kwargs)
            return response
        except botocore.exceptions.ClientError as exc:
            status = "error"
            error_code = exc.response.get("Error", {}).get("Code", "UnknownError")
            if error_code == "ThrottlingException":
                status = "throttled"
            raise
        finally:
            latency_ms = (time.monotonic() - start) * 1000
            request_id = (
                response.get("ResponseMetadata", {}).get("RequestId")
                if response
                else None
            )
            event = self._build_event(
                invocation_type="invoke_model",
                model_id=model_id,
                attribution=attribution,
                response=response,
                latency_ms=latency_ms,
                status=status,
                error_code=error_code,
                request_id=request_id,
            )
            self._wrapper.emit_event(event)

    def __getattr__(self, name: str):
        """Pass through any other client methods uninstrumented."""
        return getattr(self._client, name)


# ──────────────────────────────────────────────
# Bedrock Agents step interceptor
# ──────────────────────────────────────────────

class AgentStepInterceptor:
    """
    Captures token usage from Bedrock Agent invocation responses.

    Bedrock Agents don't expose per-step token counts directly via the
    InvokeAgent API. This interceptor wraps invoke_agent and parses the
    response stream to accumulate token usage across steps.

    Usage:
        interceptor = AgentStepInterceptor(wrapper, agent_id="XXXXXXXXXX")
        response = interceptor.invoke_agent(
            agentId="XXXXXXXXXX",
            agentAliasId="YYYYYYYYYY",
            sessionId="sess_123",
            inputText="What is my account balance?",
            _user_id="user_abc",
        )
    """

    def __init__(
        self,
        wrapper: BedrockCostWrapper,
        region: Optional[str] = None,
    ):
        self._wrapper = wrapper
        self._region = region or os.environ.get("AWS_DEFAULT_REGION", "us-east-1")
        self._agents_client = boto3.client(
            "bedrock-agent-runtime", region_name=self._region
        )

    def invoke_agent(self, **kwargs) -> dict:
        """
        Instrumented invoke_agent. Reads the response event stream and
        accumulates token usage from 'metadata' event chunks.
        """
        user_id = kwargs.pop("_user_id", None)
        agent_id = kwargs.get("agentId", "unknown")
        session_id = kwargs.get("sessionId")
        model_id = f"agent/{agent_id}"

        start = time.monotonic()
        input_tokens = output_tokens = 0
        status = "success"
        error_code = None

        try:
            response = self._agents_client.invoke_agent(**kwargs)
            event_stream = response.get("completion", [])

            # Drain the stream — must consume to get token counts
            collected_chunks = []
            for stream_event in event_stream:
                collected_chunks.append(stream_event)

                # Bedrock Agents emit a 'metadata' event per step with usage
                if "metadata" in stream_event:
                    usage = stream_event["metadata"].get("usage", {})
                    input_tokens += usage.get("inputTokens", 0)
                    output_tokens += usage.get("outputTokens", 0)

            # Rebuild a response object with the drained stream
            response["_collected_chunks"] = collected_chunks
            return response

        except botocore.exceptions.ClientError as exc:
            status = "error"
            error_code = exc.response.get("Error", {}).get("Code", "UnknownError")
            raise
        finally:
            latency_ms = (time.monotonic() - start) * 1000
            event = {
                "event_id": str(uuid.uuid4()),
                "schema_version": 1,
                "timestamp": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
                "region": self._region,
                "account_id": self._wrapper._get_account_id(),
                "model_id": model_id,
                "invocation_type": "agent_step",
                "user_id": user_id,
                "agent_id": agent_id,
                "session_id": session_id,
                "application_id": self._wrapper.application_id,
                "request_id": None,
                "input_tokens": input_tokens,
                "output_tokens": output_tokens,
                "cache_read_tokens": 0,
                "cache_write_tokens": 0,
                "latency_ms": int(latency_ms),
                "status": status,
                "error_code": error_code,
                "source": "wrapper",
            }
            self._wrapper.emit_event(event)
